src/cockpit/superuser.py

```diff
@@ -0,0 +1,215 @@
+    startup: Optional[SuperuserStartup]
+    peer: Optional[Peer]
+
+    # D-Bus signals
+    prompt = bus.Interface.Signal('s', 's', 's', 'b', 's')  # message, prompt, default, echo, error
+
+    # D-Bus properties
+    bridges = bus.Interface.Property('as', value=[])
+    current = bus.Interface.Property('s', value='none')
+
+    # RoutingRule
+    def apply_rule(self, options: Dict[str, object]) -> Optional[Peer]:
+        if not options.get('superuser') or self.current == 'root':
+            # superuser not requested, or already superuser?  Next rule.
+            return None
+        elif self.peer:
+            # superuser requested and active?  Return it.
+            return self.peer
+        else:
+            # superuser requested, but not active?  That's an error.
+            raise RoutingError('access-denied')
+
+    # PeerStateListener hooks
+    def peer_state_changed(self, peer: Peer, event: str, exc: Optional[Exception] = None) -> None:
+        logger.debug('Peer %s state changed -> %s', peer.name, event)
+        if event == 'connected':
+            self.current = 'init'
+
+        elif event == 'init':
+            self.peer = peer
+            self.current = peer.name
+            if self.startup is not None:
+                self.startup.success(self)
+                self.startup = None
+
+        elif event == 'closed':
+            self.current = 'none'
+            self.peer = None
+
+            if self.startup is not None:
+                self.startup.failed(self, exc or ConnectionResetError('connection lost'))
+                self.startup = None
+
+    def peer_authorization_request(self, peer: Peer, prompt: str, echo: bool) -> None:
+        if self.startup:
+            self.startup.auth(self, prompt, echo)
+
+    def __init__(self, router: Router, privileged: bool = False):
+        super().__init__(router)
+
+        self.bridges = list(SUPERUSER_BRIDGES)
+        self.peer = None
+        self.startup = None
+
+        if privileged or os.getuid() == 0:
+            self.current = 'root'
+
+    def go(self, startup: SuperuserStartup, name: str) -> None:
+        assert self.current == 'none'
+        assert self.peer is None
+        assert self.startup is None
+
+        args, env = SUPERUSER_BRIDGES[name]
+        startup.peer = Peer(self.router, name, self)
+        startup.peer.spawn(args, env)
+        self.startup = startup
+
+    def init(self, params: Dict[str, object]) -> None:
+        name = params.get('id')
+        if not isinstance(name, str) or name == 'any':
+            name = 'sudo'
+        self.go(ControlMessageStartup(), name)
+
+    # D-Bus methods
+    @bus.Interface.Method(in_types=['s'])
+    async def start(self, name: str) -> None:
+        try:
+            startup = DBusStartup()
+            self.go(startup, name)
+            await startup.wait()
+        except AssertionError as exc:
```

Comment:
I feel this is better done with a direct check: `if self.current != 'none':`

Reply:
"Ask forgiveness, not permission"

Reply:
But indeed, for assertions, I could be convinced of the other way.

Reply:
Done.